### PR TITLE
feat: cherry-pick improvements from `gogo/protobuf`

### PR DESCRIPTION
**Pull request by julienrbrt**
Cherry-pick:

- https://github.com/gogo/protobuf/pull/723
- https://github.com/gogo/protobuf/pull/710
- https://github.com/gogo/protobuf/pull/682

I tried to fix the `make regenerate` without too much success (although I have not looked at it very long).
@robert-zaremba has it ever worked?

### CODE REVIEW

**Comment by robert-zaremba** [comment]
@julienrbrt yes it worked. I was able to generate all the files. But I had to make many changes. See my commits.

File: LICENSE [hunk 1]
```diff
@@ -1,3 +1,4 @@
+Copyright (c) 2022, The Cosmos SDK Authors. All rights reserved.
```
Comment (robert-zaremba) [inline LICENSE:1]: :+1:

File: conformance/Makefile [hunk 2]
```diff
@@ -49,11 +49,11 @@ test:
 	./test.sh $(PROTOBUF_ROOT)
 
 regenerate:
-	protoc-min-version --version="3.0.0" --proto_path=$(GOPATH)/src:$(GOPATH)/src/github.com/cosmos/gogoproto/protobuf:. --gogo_out=\
+	protoc-min-version --version="3.0.0" --proto_path=../protobuf:. --gogo_out=\
```
Comment (robert-zaremba) [inline conformance/Makefile:52]: why changing this?

File: conformance/Makefile [hunk 3]
```diff
@@ -49,11 +49,11 @@ test:
...
 	Mgoogle/protobuf/any.proto=github.com/cosmos/gogoproto/types,\
 	Mgoogle/protobuf/duration.proto=github.com/cosmos/gogoproto/types,\
 	Mgoogle/protobuf/struct.proto=github.com/cosmos/gogoproto/types,\
 	Mgoogle/protobuf/timestamp.proto=github.com/cosmos/gogoproto/types,\
 	Mgoogle/protobuf/wrappers.proto=github.com/cosmos/gogoproto/types,\
 	Mgoogle/protobuf/field_mask.proto=github.com/cosmos/gogoproto/types\
-	:. ./internal/conformance_proto/conformance.proto
+	:. ./internal/conformance_proto/conformance.proto -I ../gogoproto -I ../protobuf
```
Comment (robert-zaremba) [inline conformance/Makefile:59]:
```suggestion
	:. ./internal/conformance_proto/conformance.proto\
	 -I ../gogoproto -I ../protobuf
```

**Review by robert-zaremba — APPROVED** [review]
utACK, left few comments

File: conformance/Makefile (hunk 2 above)
Comment (julienrbrt) [reply conformance/Makefile:52]: Because we don't use the `GOPATH`

**Pull request merged**